### ISSUE TITLE
[release][xgboost] ft_small_elastic fails

### ISSUE THREAD

**Issue by wuisawesome** [ray-project/ray#13949]
### What is the problem?

`ft_small_elastic` fails with the following error:

```
2021-02-05 15:21:41,820	WARNING main.py:1080 -- A Ray actor died during training. Trying to continue training on the remaining actors. This will use 4 existing actors and start 0 new actors. Sleeping for 10 seconds for cleanup.
[2m[36m(pid=3919)[0m [15:21:41] WARNING: ../rabit/src/allreduce_base.cc:446: failed in ReconnectLink [15:21:41] ../rabit/include/rabit/internal/utils.h:90: ReConnectLink failure 4
[2m[36m(pid=3919)[0m [15:21:43] WARNING: ../rabit/src/allreduce_base.cc:240: Retry connect to ip(retry time 1): [172.31.16.26]
[2m[36m(pid=3919)[0m 
[2m[36m(pid=3919)[0m [15:21:45] WARNING: ../rabit/src/allreduce_base.cc:240: Retry connect to ip(retry time 2): [172.31.16.26]
[2m[36m(pid=3919)[0m 
[2m[36m(pid=3919)[0m [15:21:47] WARNING: ../rabit/src/allreduce_base.cc:237: Connect to (failed): [172.31.16.26]
[2m[36m(pid=3919)[0m 
[2m[36m(pid=3919)[0m [15:21:47] WARNING: ../rabit/src/allreduce_base.cc:139: Failed to shutdown due to[15:21:47] ../rabit/include/rabit/internal/utils.h:114: Socket Connect Error:Connection refused
[2m[36m(pid=3919)[0m [15:21:49] WARNING: ../rabit/src/allreduce_base.cc:240: Retry connect to ip(retry time 3): [172.31.16.26]
[2m[36m(pid=3919)[0m 
2021-02-05 15:21:51,838	INFO main.py:729 -- [RayXGBoost] Created 0 new actors (4 total actors). Waiting until actors are ready for training.
2021-02-05 15:21:51,944	INFO main.py:766 -- [RayXGBoost] Starting XGBoost training.
2021-02-05 15:21:52,853	INFO elastic.py:155 -- Actor status: 4 alive, 0 dead (4 total)
[2m[36m(pid=3919)[0m [15:21:55] WARNING: ../rabit/src/allreduce_base.cc:240: Retry connect to ip(retry time 4): [172.31.16.26]
[2m[36m(pid=3919)[0m 
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 838, in _train
    ray.get(ready)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 1456, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): [36mray::RayXGBoostActor.train()[39m (pid=1884, ip=172.31.28.117)
  File "python/ray/_raylet.pyx", line 442, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 459, in ray._raylet.execute_task
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/serialization.py", line 242, in deserialize_objects
    assert self.get_outer_object_ref() is None
AssertionError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 1049, in train
    **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 864, in _train
    raise RayActorError from exc
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task. Check python-core-worker-*.log files for more information.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ray/ft_small_elastic.py", line 54, in <module>
    id="second_fail", state=failure_state, ranks=[0], iteration=34)
  File "/home/ray/_train.py", line 87, in train_ray
    **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 1107, in train
    ) from exc
RuntimeError: A Ray actor died during training and the maximum number of retries (3) is exhausted.
[2m[36m(pid=3919)[0m [15:21:57] WARNING: ../rabit/src/allreduce_base.cc:446: failed in ReconnectLink [15:21:57] ../rabit/include/rabit/internal/utils.h:90: ReConnectLink failure 4
```

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
ray rsync-up cluster_cpu_small-autoscaler.yaml workloads/ft_small_non_elastic.py ~/ft_small_non_elastic.py
ray submit cluster_cpu_small-autoscaler.yaml workloads/ft_small_elastic.py
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by krfricke** [comment]
I just ran the test successfully:

```
(pid=3387, ip=172.31.29.47) [05:34:44] task [xgboost.ray]:140275581341776 got new rank 3
2021-02-08 05:35:12,412 INFO main.py:919 -- [RayXGBoost] Finished XGBoost training on training data with total N=2,000,000.
TRAIN TIME TAKEN: 64.58 seconds
Final training error: 0.4183
PASSED.
```

Two notes here. As described in the instructions, this test is flaky, and it sometimes fails in two different places:

1) Sometimes additional training actors are killed. *This seems to be a problem with xgboost_ray* and we are actively working on resolving this.

2) Data loading is sometimes so fast that the test does not see any training with fewer than 4 actors, because the killed actor is re-integrated into training right away. This is just an issue of reporting the correct response. I fixed this in the PR for the cluster yamls.

Since 1 is a problem of xgboost_ray, not ray itself, the release of the core library shouldn't be blocked by this.
